File: project/drivers/driver/php/tests/MetadataRecursiveSchemaTest.php
<?php
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

require_once __DIR__ . '/bootstrap.php';

use PHPUnit\Framework\TestCase;
use ScratchBird\PDO\Metadata;

final class MetadataRecursiveSchemaTest extends TestCase
{
    public function testExpandSchemaMetadataRowsSupportsDatabaseDefaultBranchStyleRows(): void
    {
        $rows = [
            [
                'schema_id' => 11,
                'TABLE_SCHEM' => 'database.default.users',
                'TABLE_CATALOG' => 'database',
            ],
            [
                'schema_id' => 12,
                'TABLE_SCHEM' => 'database.default.audit',
                'TABLE_CATALOG' => 'database',
            ],
        ];

        $expanded = Metadata::expandSchemaMetadataRows($rows);

        $this->assertSame(
            ['database', 'database.default', 'database.default.users', 'database.default.audit'],
            $this->collectSchemaValues($expanded, 'TABLE_SCHEM')
        );
        $this->assertSame([null, null, 11, 12], $this->collectColumnValues($expanded, 'schema_id'));
    }

    public function testListMetadataSchemaPathsExpandsDottedSchemaParents(): void
    {
        $rows = [
            $this->schemaRow('users.alice.dev'),
            $this->schemaRow('sys'),
            $this->schemaRow('users.bob.dev'),
            $this->schemaRow('users.bob.dev'),
            $this->schemaRow('users..bob.dev'),
            $this->schemaRow(''),
        ];

        $expanded = Metadata::listMetadataSchemaPaths($rows, true);
        $this->assertSame(
            [
                'users',
                'users.alice',
                'users.alice.dev',
                'sys',
                'users.bob',
                'users.bob.dev',
            ],
            $expanded
        );
    }

    public function testBuildMetadataSchemaTreeNoDuplicatesWithinSameParent(): void
    {
        $rows = [
            $this->schemaRow('users.bob.dev'),
            $this->schemaRow('users.bob.dev'),
            $this->schemaRow('users.bob.prod'),
        ];

        $tree = Metadata::buildMetadataSchemaTree($rows);
        $bob = $this->findNodeByPath($tree['schemas'] ?? [], 'users.bob');
        $this->assertNotNull($bob);
        $this->assertCount(2, $bob['children']);
        $this->assertSame(
            ['users.bob.dev', 'users.bob.prod'],
            array_values(array_map(static fn (array $node): string => $node['path'], $bob['children']))
        );
        $this->assertSame(
            1,
            count(array_filter($bob['children'], static fn (array $node): bool => $node['name'] === 'dev'))
        );
    }

    public function testBuildMetadataSchemaTreeAllowsSameLeafUnderDifferentParents(): void
    {
        $rows = [
            $this->schemaRow('users.alice.dev'),
            $this->schemaRow('users.bob.dev'),
        ];

        $tree = Metadata::buildMetadataSchemaTree($rows, true, 'demo');
        $aliceDev = $this->findNodeByPath($tree['schemas'] ?? [], 'users.alice.dev');
        $bobDev = $this->findNodeByPath($tree['schemas'] ?? [], 'users.bob.dev');

        $this->assertSame('demo', $tree['database']);
        $this->assertNotNull($aliceDev);
        $this->assertNotNull($bobDev);
        $this->assertSame('dev', $aliceDev['name']);
        $this->assertSame('dev', $bobDev['name']);
        $this->assertNotSame($aliceDev['path'], $bobDev['path']);
    }

    private function schemaRow(string $schema): array
    {
        return ['schema_name' => $schema];
    }

    /**
     * @param array<int, array<string, mixed>> $rows
     * @return array<int, mixed>
     */
    private function collectColumnValues(array $rows, string $column): array
    {
        return array_values(
            array_map(
                static fn (array $row): mixed => $row[$column] ?? null,
                $rows
            )
        );
    }

    /**
     * @param array<int, array<string, mixed>> $rows
     * @return array<int, string>
     */
    private function collectSchemaValues(array $rows, string $key): array
    {
        $out = [];
        foreach ($rows as $row) {
            if (!isset($row[$key]) || !is_string($row[$key])) {
                continue;
            }
            $out[] = $row[$key];
        }
        return $out;
    }

    /**
     * @param array<int, array<string, mixed>> $nodes
     * @return array<string, mixed>|null
     */
    private function findNodeByPath(array $nodes, string $path): ?array
    {
        foreach ($nodes as $node) {
            if (($node['path'] ?? null) === $path) {
                return $node;
            }
            if (!isset($node['children']) || !is_array($node['children'])) {
                continue;
            }
            $child = $this->findNodeByPath($node['children'], $path);
            if ($child !== null) {
                return $child;
            }
        }
        return null;
    }
}

File: packaging/2026.07.03/drivers/dart/examples/test/metadata_recursive_schema_test.dart
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

import 'package:scratchbird/scratchbird.dart';
import 'package:test/test.dart';

void main() {
  group('metadata schema shaping', () {
    test('expands database-default branch style metadata rows', () {
      final rows = <Map<String, dynamic>>[
        {
          'TABLE_CATALOG': 'demo_db',
          'TABLE_SCHEM': 'users.alice.dev',
          'schema_id': 1,
          'owner_id': 7,
        },
        {
          'TABLE_CATALOG': 'demo_db',
          'TABLE_SCHEM': 'sys',
          'schema_id': 2,
          'owner_id': 7,
        },
      ];

      final expanded = expandSchemaMetadataRows(rows, expandParents: true);

      expect(
        expanded.map((row) => row['TABLE_SCHEM']),
        equals(<String>['users', 'users.alice', 'users.alice.dev', 'sys']),
      );
      expect(expanded[0]['TABLE_CATALOG'], equals('demo_db'));
      expect(expanded[1]['TABLE_CATALOG'], equals('demo_db'));
      expect(expanded[0]['schema_id'], isNull);
      expect(expanded[2]['schema_id'], equals(1));
      expect(expanded[3]['schema_id'], equals(2));
    });

    test('supports dotted parent expansion for schema paths', () {
      final paths = listMetadataSchemaPaths(
        <Map<String, dynamic>>[
          {'schema_name': 'analytics.prod'},
          {'schema_name': 'users.alice.dev'},
          {'schema_name': 'users.alice.dev'},
          {'schema_name': 'users..bob.dev'},
        ],
        expandParents: true,
      );

      expect(
        paths,
        equals(<String>[
          'analytics',
          'analytics.prod',
          'users',
          'users.alice',
          'users.alice.dev',
          'users.bob',
          'users.bob.dev',
        ]),
      );
    });

    test('enforces uniqueness within same parent', () {
      final tree = buildMetadataSchemaTree(
        <Map<String, dynamic>>[
          {'schema_name': 'users.bob.dev'},
          {'schema_name': 'users.bob.dev'},
          {'schema_name': 'users.bob.qa'},
        ],
      );

      final users = _findNodeByPath(tree.schemas, 'users');
      final bob = _findNodeByPath(users!.children, 'users.bob');
      expect(bob, isNotNull);
      expect(
        bob!.children.map((child) => child.path).toList(),
        equals(<String>['users.bob.dev', 'users.bob.qa']),
      );
      expect(
          bob.children.where((child) => child.path == 'users.bob.dev').length,
          equals(1));
    });

    test('allows same leaf name under different parents', () {
      final tree = buildMetadataSchemaTree(
        <Map<String, dynamic>>[
          {'schema_name': 'users.alice.dev'},
          {'schema_name': 'users.bob.dev'},
        ],
        database: 'demo_db',
      );

      expect(tree.database, equals('demo_db'));
      expect(
        tree.schemas.map((node) => node.path).toList(),
        equals(<String>['users']),
      );

      final users = _findNodeByPath(tree.schemas, 'users');
      final alice = _findNodeByPath(users!.children, 'users.alice');
      final bob = _findNodeByPath(users.children, 'users.bob');
      final aliceDev = _findNodeByPath(alice!.children, 'users.alice.dev');
      final bobDev = _findNodeByPath(bob!.children, 'users.bob.dev');

      expect(aliceDev, isNotNull);
      expect(bobDev, isNotNull);
      expect(aliceDev, isNot(same(bobDev)));
      expect(aliceDev!.name, equals('dev'));
      expect(bobDev!.name, equals('dev'));
    });
  });
}

MetadataSchemaTreeNode? _findNodeByPath(
  List<MetadataSchemaTreeNode> nodes,
  String path,
) {
  for (final node in nodes) {
    if (node.path == path) {
      return node;
    }
  }
  return null;
}
